src/eng_back.c

```diff
@@ -257,7 +257,43 @@ static int ENGINE_CTX_ctrl_set_vlog(ENGINE_CTX *ctx, void *cb)
 	return 1;
 }
 
-int ENGINE_CTX_ctrl(ENGINE_CTX *ctx, int cmd, long i, void *p, void (*f)(void))
+static int ENGINE_CTX_keygen(ENGINE_CTX *ctx, void *p)
+{
+	int rv = 0;
+	PKCS11_KGEN_ATTRS *kg_attrs = p;
+	PKCS11_SLOT* found_slot = NULL;
+
+	if (kg_attrs == NULL)
+		return 0;
+
+	/* Delayed libp11 initialization */
+	if (UTIL_CTX_init_libp11(ctx->util_ctx)) {
+		ENGerr(ENG_F_CTX_LOAD_OBJECT, ENG_R_INVALID_PARAMETER);
+		return 0;
+	}
+
+	found_slot = UTIL_CTX_find_token(ctx->util_ctx, kg_attrs->token_label);
+	if (!found_slot)
+		return 0;
+
+	/* Try with logging in */
+	ERR_clear_error();
+	if (found_slot->token->loginRequired && UTIL_CTX_login(ctx->util_ctx,
```

Comment:
If token login is required, UTIL_CTX_login() must be called. However, try to generate a key pair.